File: leetcode2/2medium/김대원/Q2587.java
// 1. 아이디어 : 누적합을 생성하고 양수 판별
// 2. 시간복잡도 : O(log N)
// 3. 자료구조/알고리즘 : 누적합

class Solution {
    public int maxScore(int[] nums) {
        Arrays.sort(nums);
        int n = nums.length;
        long[] prefixSum = new long[n];
        prefixSum[0] = nums[n - 1];
        int answer = 0;
        for (int i = 1; i < n; i++) {
            prefixSum[i] = prefixSum[i - 1] + nums[n - i - 1];
            if (prefixSum[i] > 0) answer++;
        }

        return prefixSum[0] > 0 ? answer + 1 : answer;
    }
}

File: leetcode2/2medium/강병호/Q.2587.py
<<<<<<< HEAD
class Leetcode.이재훈.이재훈.Solution:
=======
class Leetcode.이재훈.Solution:
>>>>>>> ab7f46f73d8f18bcd38c5cf06ecabae64f53a424
    def maxScore(self, nums: list[int]) -> int:
        sorted_nums = sorted(nums, reverse=True)

        result = 0
        prefix = 0

        for data in sorted_nums:
            result += data
            if result <= 0:
                break
            else:
                prefix += 1
        
        return prefix
    

if __name__ == "__main__":
<<<<<<< HEAD
    solution = Leetcode.이재훈.이재훈.Solution()
=======
    solution = Leetcode.이재훈.Solution()
>>>>>>> ab7f46f73d8f18bcd38c5cf06ecabae64f53a424

    print(solution.maxScore([2, -1, 0, 1, -3, 3, -3]))
    print(solution.maxScore([-2, -3, 0]))